<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Rewriting systems, bytecode soups, and algorithmic chemistries</title>
<meta name="description" content="">
<meta name="author" content="Graham Wakefield">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
<link rel="stylesheet" href="css/basic.css" type="text/css" />
<link rel="stylesheet" href="css/github.css" type="text/css" />
<style>
header {
	background-color:#f5f5f5;
	font-size: 75%;
	padding: 0.5em;
}

footer {
	background-color:#f5f5f5;
	font-size: 75%;
	padding: 0.5em;
}
</style>
</head>

<body class="centremaxwidth960">
<header><a href="index.html">Home</a></header>
<ul>
<li><a href="#evo-devo">Evo Devo</a><ul>
<li><a href="#rewriting-systems-bytecode-soups-and-algorithmic-chemistries">Rewriting systems, bytecode soups, and algorithmic chemistries</a></li>
</ul>
</li>
<li><a href="#development-and-l-systems">Development and L-systems</a></li>
<li><a href="#rewriting-systems">Rewriting systems</a><ul>
<li><a href="#turtle-graphics">Turtle graphics</a></li>
<li><a href="#bracketed-systems">Bracketed systems</a></li>
<li><a href="#context-sensitive-formal-languages">Context-sensitive formal languages</a></li>
<li><a href="#parametric-systems">Parametric systems</a></li>
<li><a href="#stochastic-systems">Stochastic systems</a></li>
</ul>
</li>
<li><a href="#evo-devo-1">Evo-Devo</a></li>
<li><a href="#bytecode-ecosystems">Bytecode ecosystems</a></li>
<li><a href="#thomas-rays-tierra">Thomas Ray&#39;s Tierra</a><ul>
<li><a href="#darwinian-os-components">Darwinian OS components</a></li>
<li><a href="#more-information">More information:</a></li>
</ul>
</li>
<li><a href="#artificial-chemistries">Artificial chemistries</a></li>
<li><a href="#walter-fontanas-algorithmic-chemistry">Walter Fontana&#39;s Algorithmic Chemistry</a><ul>
<li><a href="#primitives">Primitives</a></li>
</ul>
</li>
<li><a href="#further-afield">Further afield</a></li>
<li><a href="#general-properties-of-artificial-chemistries">General properties of artificial chemistries</a></li>
</ul>

<!--

- branching system from Hry's work in 5010
- concat style languages
- ANNs! 

-->


<h2 id="evo-devo">Evo Devo</h2>
<p>In some systems the developmental process is little more than a trivial mapping, but this misses an entire and fascinating source of diversity. Incorporating more complex developmental models can lead to geometric variations that are not stored as simple parameters, to repeated segments and recursive structures, to symmetries, and to the re-application of common toolboxes toward a widely differing set of purposes -- all things that are evident in biological evolution. </p>
<p>Many systems represent genetic information as a sequence of data, such as a string of characters or binary digits, just as our DNA is a single string. But this string winds upon itself multiple times, in complex ways, to create more subtle physical macrostructures; and also applies functionally to itself as one gene can turn another gene on or off in a complex <a href="https://en.wikipedia.org/wiki/Gene_regulatory_network"><strong>regulatory network</strong></a>. That is, although it is a string at the lowest level, at higher structural and functional levels it operates like a tree or graph. (By analogy, although the <em>text</em> of a computer program is a linear string of characters, the grammar produces a tree of re-usable functions and so on, and the data/control flow it describes are more complex graphs.) Accordingly some artificial evolutionary systems also use more complex structures for genetic representations, including trees and graphs.  </p>
<p><strong>Looking back at our code-based genetics in the Biomorphs, an interesting challenge could be to extend this into more complex <em>rewriting systems</em>, such as L-systems, by embedding <em>production rules</em>, to create evolutionary-developmental systems! I.e., the genome isn&#39;t necessarily interpreted immediately into graphics, but instead/also into the production of more code for interpreting.</strong> </p>
<p><img src="https://www.karlsims.com/creatures-3swimmers-200.jpg" alt="sims"></p>
<p>This point is important to consider. Rather than forcing symmetry with turtle-spawning operators in the interpreter, one could embed codes that duplicate and transform subsets of other substrings of the same code. The interpreter becomes simpler, but the code becomes more expressive. Such a developmental approach can be taken quite far. Karl Sims&#39; evolved virtual creatures had a genotype that encoded a LISP function (the developmental system), which when run would produce a body shape (the skeletal system), but also produced an executable function for behaviour (the neural system). That is, the genetic code produces a developmental program that produces a neuro-physiological program. More generally, since we are evolving code, we may want to the genome to produce re-usable subroutines that can be used multiple times within the main phenotype program. </p>
<p>Taking this further, the developmental program could also be gradual, meaning we can model the increase in complexity of an organism from embryo to adult through the gradual rewriting of its various genetic, physiological, and neural systems.</p>
<hr>
<h1 id="rewriting-systems-bytecode-soups-and-algorithmic-chemistries">Rewriting systems, bytecode soups, and algorithmic chemistries</h1>
<h2 id="development-and-l-systems">Development and L-systems</h2>
<iframe src="https://player.vimeo.com/video/93056665" width="500" height="281" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>

<p>The structure of a multi-cellular organisms develops in a gradual process from a single fertilized cell (the <em>zygote</em>) through progressive steps of cell duplication and specialization in a process called <em>development</em>; an active area of scientific research today. We know that each cell responds to chemical signals from its neighbors and other environmental conditions in order to know how to <em>differentiate</em> (to become a specific kind of cell in the body), according to the processes of the genome. The zygote does not contain a blueprint for the final organism, simply genotypic instructions for responding to environmental stimuli, which happen (in most cases) to produce the fully developed phenotype. </p>
<p>A compact developmental representation can result in a more complex developed organism by relying on the self-organizing tendencies of the chemical substrate (such as the chemical pattern formation we saw with reaction-diffusion systems), the creative re-use of instructions (leading to the <em>modularity</em> and <em>symmetry</em> widely found in biology), and careful control over the ordering, when and for how long particular instructions operate (<em>heterochrony</em>). Development is clearly <em>parallel</em>, <em>decentralized</em>, granting some <em>robustness</em>, and yet it is also <em>context dependent</em>. It is also <em>self-limiting</em>, such that most developmental processes halt after the embryonic phase (though some may continue through the organism&#39;s lifetime, such as the <em>regeneration</em> of lost limbs).</p>
<p>In 1968 biologist Astrid Lindenmeyer proposed to model several aspects of developmental systems (cell division, cell differentiation and cell death) using rewriting systems on symbolic strings. The analogy is that each symbol represents a cell in a particular state, and the rules of transformation represent the processes of differentiating from one state to another, dividing into two cells, or of programmed cell death.</p>
<p>(Cellular automata can also be used to model developmental processes, and arguably langton&#39;s reproducing loop does so, but rewriting systems offer a very simple and abstract model that can support growth. Sims&#39; evolving creatures used a substitution method closely equivalent rewriting to generate their body structures, for example.)</p>
<p><img src="https://blog.adafruit.com/wp-content/uploads/2013/10/NervousSystemHyphaeLampClose-2.jpg" alt="nervous system lamp"></p>
<blockquote>
<p>Above is an example of a 3D-printed lampshade whose form is the product of a developmental simulation, by the <a href="http://n-e-r-v-o-u-s.com">Nervous System design studio.</a></p>
</blockquote>
<h2 id="rewriting-systems">Rewriting systems</h2>
<p>A rewriting system defines rules for the transformation of structures, typically strings of symbols. It is closely related to the formal notion of grammar. A system comprises:</p>
<ul>
<li>A set of possible symbols (A), called the <em>alphabet</em></li>
<li>An initial symbol (S), called the <em>axiom</em></li>
<li>A set (P) of <em>rewriting</em> or <em>production rules</em> to specify how each symbol is transformed to other symbols, at each production step. A rule converts a <em>predecessor</em> symbol into <em>successor</em> symbols.</li>
<li>A process of applying these rules in parallel, step by step until no more productions can be applied.</li>
</ul>
<p>For example:</p>
<ul>
<li><strong>alphabet</strong>: <code>{ &quot;B&quot;, &quot;F&quot;, &quot;+&quot;, &quot;-&quot;, &quot;[&quot;, &quot;]&quot; }</code>, </li>
<li><strong>rules</strong>: <ul>
<li><code>F -&gt; FF</code> </li>
<li><code>B -&gt; F[-B]+B</code></li>
</ul>
</li>
<li><strong>start symbol</strong>: <code>B</code> </li>
</ul>
<p>Then a sequence of productions could proceed as follows:</p>
<pre><code class="language-js">B                        
<span class="hljs-comment">// apply rule 1:</span>
F[-B]+B                    
<span class="hljs-comment">// apply rule 1 (twice), and rule 2:</span>
FF[-F[-B]+B]+F[-B]+B        
<span class="hljs-comment">// apply rule 1 three times, rule 1 four times:</span>
FFFF[-FF[-F[-B]+B]+F[-B]+B]+FF[-F[-B]+B]+F[-B]+B        
<span class="hljs-comment">// etc.</span></code></pre>
<p>Notice that any symbol that is not a predecessor of a production rule is passed through unchanged (it is not removed). </p>
<h3 id="turtle-graphics">Turtle graphics</h3>
<p>One of the simplest ways to use rewriting sytems for art and design is to interpret the produced strings as instructions for another program. The classic example is using them as instructions for a &quot;turtle graphics&quot; interpreter. </p>
<p>For example, using the following system:</p>
<pre><code class="language-javascript">    <span class="hljs-comment">// just one rule: replace &quot;F&quot; with &quot;F+F--F+F&quot;</span>
    rules[<span class="hljs-string">&quot;F&quot;</span>] = <span class="hljs-string">&quot;F+F--F+F&quot;</span>
    <span class="hljs-comment">// start with:</span>
    S = <span class="hljs-string">&quot;F&quot;</span></code></pre>
<p>If we interpret the &quot;F&quot; symbol to mean &quot;move forward&quot;, and the &quot;+&quot; and &quot;-&quot; symbols to mean turn left and right by 60 degrees, then each successive application of this rule generates a successive iteration of the Koch curve fractal:</p>
<p><img src="img/koch.jpg" alt="Iterations of the Koch curve fractal"></p>
<p><a href="http://codepen.io/grrrwaaa/pen/xVVpea?editors=001">See the Koch curve example in the lab editor</a>.</p>
<h3 id="bracketed-systems">Bracketed systems</h3>
<p>By adding push <code>&quot;[&quot;</code> and pop <code>&quot;]&quot;</code> symbols to save/restore graphics state (position, orientation etc.), the graphics interpreter can render branched structures such as trees and ferns. The result is further improved by reducing the length of each line according to the bracketed recursion depth, and embellishing with more context-sensitivity:</p>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/7/74/Dragon_trees.jpg" alt="trees"></p>
<p><a href="../code/lsystem_max.zip">See the example Max/MSP/Jitter patch here</a></p>
<h3 id="context-sensitive-formal-languages">Context-sensitive formal languages</h3>
<p>An alphabet and production rules specify a formal language. In 1959 Noam Chomsky identified a hierarchy of four types of formal languages. They rest on the concept of <em>terminal</em> and <em>non-terminal</em> symbols, where a symbol is terminal if it cannot be further rewritten. Nonterminal symbols, which can be further rewritten, are normally indicated using upper case letters (A, B, C) and terminal symbols by lowercase letters (a, b, c). </p>
<p>Each set in the hierarchy contains those above it:</p>
<ol>
<li><strong>Regular (Type-3):</strong> the predecessors of rules are a single nonterminal, the successors are a single terminal possibly followed by a single nonterminal, e.g. A -&gt; Ab or A -&gt; B.</li>
<li><strong>Context-free (Type-2):</strong> the predecessors are a single nonterminal, and the sucessors can be any string of terminals and nonterminals. </li>
<li><strong>Context-sensitive (Type-1):</strong> the predecessors contain multiple symbols (including at least one nonterminal), and the sucessor is a copy with the nonterminal replaced by an arbitrary string of nonterminals. E.g. aBc -&gt; aBbc, Ab -&gt; aaab, abcD -&gt; abcF, etc. They are therefore akin to Type-2 rules except with additional conditions regarding neighboring symbols (which remain unchanged).</li>
<li><strong>Unrestricted (Type-0):</strong> predecessors and successors can both be any finite sequence of symbols without restriction.</li>
</ol>
<p>So far we have considered only regular and context-free formal languages, since we have been restricted to single nonterminal symbols on the left-hand side (predecessors). Clearly, as a biological model (where the string is the phenotype and the production rules are genotype), our system should be context sensitive. Context would imply environmental conditions as well as neighbor cell signaling. </p>
<p>In order to support context-sensitive languages, we will need more powerful pattern-matching capabilities, and need extra care to handle ambiguous situations. E.g. we must be careful to consider longer patterns before shorter ones, to ensure that e.g. <code>aba</code> matches a rule <code>aba -&gt; abba</code> rather than the rule <code>a -&gt; b</code>. That is, our set of rules must now form a list (rather than a set), which is ordered by <em>priority</em>. And to match classes of structures, we may turn to using <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions">regular expressions</a> or more powerful <a href="https://en.wikipedia.org/wiki/Parsing_expression_grammar">parsing expression grammars</a> as the pattern and pattern-matching features.</p>
<p>The lab toolkit includes the <a href="http://pegjs.org">peg.js</a> library, available in the variable name &quot;pegjs&quot;, which can be used to build parsers. This library also has an excellent <a href="http://pegjs.org/online">online editor</a> in which you can design grammars and test them. I have a <a href="https://worldmaking.github.io/workshop_iclc_2016/tutorial.html">parsing expression grammar tutorial designed for peg.js here</a>.</p>
<h3 id="parametric-systems">Parametric systems</h3>
<p>A <strong>parametric</strong> L-system embeds parameters (such as line length and rotation angle) into the symbols, and sucessors may modify these parameter values. For example, the rule <code>F(x) -&gt; F(x*2)</code> transforms a forward step into a forward step of twice the length.</p>
<p>A <strong>conditional parametric</strong> L-system extends this to incorporate conditions, such that the subsitution only occurs if the condition is true. For example, <code>F(x : x &lt; 10) -&gt; F(x*2)</code> transforms a forward step into a forward step of twice the length, only if the initial length was less than 10. </p>
<p>In both cases, our pattern matching system now needs to be able to handle classes of symbols (or at least, ignore anything between parentheses). </p>
<h3 id="stochastic-systems">Stochastic systems</h3>
<p>So far our systems are deterministic, and will always reproduce the same results after a fixed number of steps. To create a more natural-looking result, we can introduce non-determinism:</p>
<ul>
<li>Into the interpreter (by adding small deviations to the turtle, for example). This will remove the stark repetition, but still preserves topological symmetry.</li>
<li>Into the production rules. This can generate new topologies. </li>
</ul>
<p>A <strong>stochastic</strong> L-system does the latter: for a given non-terminal predecessor, there may be several successor strings which are selected at random (possibly according to weights). Or, if the rules are listed by priority, each rule can have a probablity of being ignored.</p>
<hr>
<h2 id="evo-devo-1">Evo-Devo</h2>
<p>L-Systems are difficult to predict and control: it is difficult to infer from an L-system definition and interpreter what the results will be, since (like cellular systems) the interactions are local but the consequences are global. Note that nature faces the same challenge: it cannot predict what changes to a genotype will result is viable or better phenotypes, and must proceed by trial and error (within the scope offered by development). Biologists exploring the mechanisms of development and generation of variation within it, and the relationship with evolutionary selection, work within the relatively new discipline of <em>evolutionary development</em> or &quot;evo-devo&quot;. </p>
<p>The role of development in the generation of variation is equally important for artificial evolution, however it has been less thoroughly explored. Two lessons from the biologists:</p>
<ul>
<li>Development is not just a compact representation of a phenotype, but can be crucial to the generation of complex systems.</li>
<li>The recent discovery that widely different species re-use the same developmental genes and mechanisms. This perhaps suggests that finding a good set of re-usable mechanisms for long-term evolvability (continuous generation of selectable novelty) is not an easy task. </li>
<li>Development is not a one-way process, under command of the genome: factors in the environment can activate or inhibit parts of the genome (and parts of a genome can activate and inhibit other parts) in a complex <em>regulatory network</em>, and the state of the environment can affect the results of actions of the genome. In some cases the genome itself may not even be fixed. In part this environment is also inherited from an organisms&#39; parents. </li>
<li>Development is ultimately working within the limits of a physical, chemical system, with entropic tendencies, self-organizing potential, noisy behavior, etc.</li>
</ul>
<p>Artificial Evo-Devo systems could be categorized as follows:</p>
<ul>
<li>The developmental process is hand-written, and evolution only works on parameters for this system. This is amounts to a simple genotype-phenotype map. </li>
<li>Mechanims of development are encoded in the genotype, but in a fixed sequence (a fixed order with a fixed number of applications). This is suitable for evolving optimal L-systems for a given pattern, for example.</li>
<li>The genotype encodes the sequence of applications of hand-written development mechanisms, as used for example in genetic programming.</li>
<li>Both the basic mechanisms and the sequence of applications are encoded in the genotype and are therefore subject to evolution. Obviously the space of exploration is much larger and less predictable, but the challenge of finding suitable primitives and representations is difficult.</li>
</ul>
<p>Whether it is parameters, sequence, mechanisms (or several) that are to be evolved, they must all be represented in the genome in forms that can be inherited, with additional mechanisms for generating variation (e.g. mutations). That is, each one of these elements must be represented symbolically as data (e.g. as strings, trees or matrices of symbols, which may be as low-level as binary form or as high-level as readable text). </p>
<hr>
<h2 id="bytecode-ecosystems">Bytecode ecosystems</h2>
<p>Perhaps the most abstract and deeply computational artificial life models are those in which the organisms <em>are</em> the code, serving as environments for each other. This approach plays most directly upon the fundamental insight that <strong>code is data</strong>. </p>
<p>We generally assume that DNA also functions as both data and behaviour, but it has sometimes been overlooked that this is a <em>material process</em>. In bytecode soups, the code is not only data, it is the material on which code acts: fragments of code can react with each other, consume and produce each other, match or subvert each other, compete with each other for CPU time or memory resources, and thus evolve. These are the most digital of the artificial life organisms.</p>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/c/c2/Core_War_PMars_Screenshot.png" alt="Core war"></p>
<p>Since at least the early 1960&#39;s people have enjoyed creating virtual machines to explore the possibility of programs competing for control of the virtual machine processors and memory. The earliest example is the game Darwin, developed in 1961, in which programs competed to stop each other from executing. Later in <em>Core War</em>, another programming game in which programs (written in an abstract assembly language called &quot;redcode&quot;) battle for control of a virtual computer, players discovered that rapid self-replication was a powerful strategy. (Virus and anti-virus systems have been modeled in these terms, although they also draw upon theories of immune systems and operate outside of protected environments.)</p>
<p>Steen Rasmussen took this idea and added genetic algorithms to evolve digital organisms in his <em>Coreworld</em> project; however complex evolution did not ensue, apparently because the assembly language was too brittle: mutations to code were far too likely to completely destroy a program. This problem of brittleness was overcome by the Tierra project.</p>
<h2 id="thomas-rays-tierra">Thomas Ray&#39;s Tierra</h2>
<p>&quot;We can think of evolution as an artist...&quot; - Tom Ray, The Tierra Video</p>
<ul>
<li><a href="http://life.ou.edu/Video/Tierra.html">The Tierra Video</a>, or here on youtube:</li>
</ul>
<iframe width="480" height="360" src="https://www.youtube.com/embed/Wl5rRGVD0QI?rel=0" frameborder="0" allowfullscreen></iframe>

<p>Tom Ray was an evolutionary biologist working in tropical rainforests for 20 years, but frustrated with the difficulty of studying evolution in the natural world, due to the restriction of using fossil records left over millions of years. So he decided to try exploring evolution in a computer, where it can operate much faster, and offer alternative models of life to study.</p>
<p>Although he didn&#39;t really know anything about how computers work, he decided to write very simple self-replicating organisms in machine language. But rather than real machine code, he created a byte code language for a virtual computer -- since he was afraid that with real machine code it might break out and infect the internet like a recent Worm virus). </p>
<p>Nobody thought it would work, but remarkably, in his first ever run of the system, starting from a simple prototype organism, these organisms improved themselves multiple times, spawned parasites and hyperparasites, immune systems, symbiotic relationships, and more. Tierra found its way into Science, Scientific American, Economist, etc. It may have been the first Von Neuman computer designed for and showing evidence of evolvability.</p>
<blockquote>
<p>The Tierra C source code creates a virtual computer and its Darwinian operating system, whose architecture has been designed in such a way that the executable machine codes are evolvable. This means that the machine code can be mutated (by flipping bits at random) or recombined (by swapping segments of code between algorithms), and the resulting code remains functional enough. Along with the C source code which generates the virtual computer, we provide programs written in the assembler code of the virtual computer, written by a human, and which do nothing more than make copies of themselves in the RAM of the virtual computer. This system results in the production of synthetic organisms based on a computer metaphor of organic life in which CPU time is the <code>energy&#39;&#39; resource and memory is the </code>material&#39;&#39; resource. Memory is organized into informational patterns that exploit CPU time for self-replication. Mutation generates new forms, and evolution proceeds by natural selection as different genotypes compete for CPU time and memory space. <a href="http://life.ou.edu/tierra/whatis.html">Adapted from &quot;What Tierra Is&quot;</a></p>
</blockquote>
<p>An organism is a program of code that occupies memory, and to be a viable organism this code must achieve self-replication, copying itself into another location in memory. One of the visualizations of Tierra represents the virtual memory visually as a sequence of bars, in which organisms are notated with specific colors.</p>
<p><img src="http://life.ou.edu/pic/almonda3.jpg" alt="Almond visualization"></p>
<blockquote>
<p>Each image represents a soup of 60,000 bytes, divided into 60 segments of 1000 bytes each. Each individual creature is represented by a colored bar, colors correspond to genome size (e.g., red = 80, yellow = 45, blue = 79). -- <a href="http://life.ou.edu/pubs/images/">As described here</a></p>
</blockquote>
<p>The environment however is volatile: </p>
<ul>
<li>From time to time bits or bytes are randomly modified as if hit by cosmic rays, causing mutations. </li>
<li>Moreover, once the memory is filled, a &quot;reaper&quot; process moves through the memory deleting programs that are old or defective (unable to run or replicate).</li>
</ul>
<p>By this combination of a population, mutation, replication with heredity, and selection pressures of viability, the entire bytecode system creates artificial evolution.</p>
<h3 id="darwinian-os-components">Darwinian OS components</h3>
<p>Tierra inherits many ideas from Core War, including circular memory, small instruction sets, instruction-level time sharing, and prevention of external access. However in other regards it differs with biological goals.</p>
<p>Effectively, the virtual computer is an example of a virtual &quot;chemical soup&quot; with unique &quot;algorithmic chemistry&quot;, in which organisms are both code and data.</p>
<ul>
<li><p>Soup (RAM): A fixed-size, circular ring, typically seeded with at least one functional organism.</p>
</li>
<li><p>Organisms: Each organism is a virtual CPU, operating in a simulated parallel fashion via the Slicer. Each CPU has its own set of 4 registers and 10 stack slots for temporary calculations (attempts to use more slots will wrap around), as well as the current instruction pointer and stack pointer. </p>
</li>
<li><p>Slicer (timesharing/threading service): A circular queue of organisms (processes). It allocates CPU time to each process in the queue in succession. Each processes&#39; time slice can be a fixed length, or a uniform random variate, determined by the genome length, etc. When a new organism is born it is added to the queue behind its mother.</p>
</li>
<li><p>Reaper: A linear queue of organisms (processes). Processes enter the bottom of the queue at birth. If the memory is full, the Reaper kills processes at the top of the queue, removing them from both Reaper and Slicer. (A killed process could have its memory zeroed, randomized, or left as-is.) Older processes generally die first. When a process generates an error, it moves one position up the Reaper queue, while a successful birth of an offspring moves the process one position down.</p>
</li>
<li><p>Memory allocator: An organism (process) is born with memory given to it by the parent. It may have read, write, or execute protection (the default is write protection only, such that other processes may not write to the block). A process may only spawn one offspring at a time, and may only reserve one memory block for each (an additional memory request before giving birth frees the offspring memory block without giving birth and allocates a new one, and instructions copied to the old block are not used). The memory allocator could be configured to locate available space by first fit, better fit, random preference, near mother&#39;s address, and suggested address. The allocator is connected to the Reaper: when a memory request cannot be fulfilled because a block of the requested size is not available, the Reaper kills processes until a block of the requested size becomes available. The size of memory requests are limited.</p>
</li>
<li><p>Volatile conditions: </p>
<ul>
<li>sporadic mutations, e.g. randomly selected memory slot is randomized</li>
<li>mutations at birth</li>
<li>mutations while copying code</li>
<li>gene splicing at birth -- crossover, insertion, deletion, possibly at or not at template boundaries (see documentation for details)</li>
<li>flaws during execution, e.g. +1 or -1 error during addition</li>
<li>sporadic killing of organisms at random</li>
</ul>
</li>
</ul>
<p>Instruction set:</p>
<p>In designing his virtual computer and its machine code, Ray asked: why can you randomly alter DNA, but you can&#39;t randomly alter a computer program? So he designed his bytecode instruction set to make it more robust:</p>
<ul>
<li><p>Tierra originally had only 32 instructions (closer to the 20 amino-acid symbols of DNA than the much larger sets common in real PC architectures.) This makes random code  far more likely to be valid.</p>
</li>
<li><p>Jump addresses described as templates, rather than fixed numbers (which would be likely to break under mutations). Template-matching is akin to the shape-matching used at microbiological levels. So, for example, the instruction sequence <code>jmp nop0 nop0 nop1</code> causes execution of the program to jump to the nearest occurrence of the instruction sequence <code>nop1 nop1 nop0</code>. The complementarity (001 finds 110) ensures that the jump will not find a copy of itself.</p>
</li>
<li><p>Apart from template-using instructions, all other instructions are atomic, requiring almost no syntax constraints. This also makes random code far more likely to be valid.</p>
</li>
<li><p>No instruction can generate an error that would crash the virtual machine. Other errors are ignored (the instruction has no effect), but may reduce organism fitness (moving down the reaper list).</p>
</li>
<li><p>pushA, pushB, pushC, pushD: copy register values onto stack</p>
</li>
<li><p>popA, popB, popC, popD: remove &amp; copy stack values onto registers</p>
</li>
<li><p>movcd, movab: copy one register to another</p>
</li>
<li><p>movii: copy from memory[A] to memory[B]</p>
</li>
<li><p>subCAB, subAAC, incA, incB, decC, zero etc.: simple arithmetic</p>
</li>
<li><p>not0, sh1: bit flip/shift</p>
</li>
<li><p>ifz: skip next instruction if C is zero</p>
</li>
<li><p>jmp, jmpb: jump to template (forward or backward)</p>
</li>
<li><p>call: push instruction pointer onto stack then jump to template</p>
</li>
<li><p>ret: jump to instruction according to stack value</p>
</li>
<li><p>adrf, adrb: search (fwd/back) for template, put address &amp; template size in register C</p>
</li>
<li><p>mal: allocate memory of size C</p>
</li>
<li><p>divide: create a child</p>
</li>
<li><p>nop0, nop1: do nothing -- these serve only as labels for template matching</p>
</li>
</ul>
<p>Ray&#39;s first ancestor organism looked like this:</p>
<pre><code>nop1    ; 110 01   0 beginning marker
nop1    ; 110 01   1 beginning marker
nop1    ; 110 01   2 beginning marker
nop1    ; 110 01   3 beginning marker
zero    ; 110 04   4 put zero in cx
not0    ; 110 02   5 put 1 in first bit of cx
shl     ; 110 03   6 shift left cx (cx = 2)
shl     ; 110 03   7 shift left cx (cx = 4)
movDC   ; 110 18   8 move cx to dx (dx = 4)
adrb    ; 110 1c   9 get (backward) address of beginning marker -&gt; ax
nop0    ; 100 00  10 complement to beginning marker
nop0    ; 100 00  11 complement to beginning marker
nop0    ; 100 00  12 complement to beginning marker
nop0    ; 100 00  13 complement to beginning marker
subAAC  ; 110 07  14 subtract cx from ax, result in ax
movBA   ; 110 19  15 move ax to bx, bx now contains start address of mother
adrf    ; 110 1d  16 get (forward) address of end marker -&gt; ax
nop0    ; 100 00  17 complement to end marker
nop0    ; 100 00  18 complement to end marker
nop0    ; 100 00  19 complement to end marker
nop1    ; 100 01  20 complement to end marker
incA    ; 110 08  21 increment ax, to include dummy instruction at end
subCAB  ; 110 06  22 subtract bx from ax to get size, result in cx
nop1    ; 110 01  23 reproduction loop marker
nop1    ; 110 01  24 reproduction loop marker
nop0    ; 110 00  25 reproduction loop marker
nop1    ; 110 01  26 reproduction loop marker
mal     ; 110 1e  27 allocate space (cx) for daughter, address to ax
call    ; 110 16  28 call template below (copy procedure)
nop0    ; 100 00  29 copy procedure complement
nop0    ; 100 00  30 copy procedure complement
nop1    ; 100 01  31 copy procedure complement
nop1    ; 100 01  32 copy procedure complement
divide  ; 110 1f  33 create independent daughter cell
jmpo    ; 110 14  34 jump to template below (reproduction loop)
nop0    ; 100 00  35 reproduction loop complement
nop0    ; 100 00  36 reproduction loop complement
nop1    ; 100 01  37 reproduction loop complement
nop0    ; 100 00  38 reproduction loop complement
ifz     ; 000 05  39 dummy instruction to separate templates
nop1    ; 110 01  40 copy procedure template
nop1    ; 110 01  41 copy procedure template
nop0    ; 110 00  42 copy procedure template
nop0    ; 110 00  43 copy procedure template
pushA   ; 110 0c  44 push ax onto stack
pushB   ; 110 0d  45 push bx onto stack
pushC   ; 110 0e  46 push cx onto stack
nop1    ; 110 01  47 copy loop template
nop0    ; 110 00  48 copy loop template
nop1    ; 110 01  49 copy loop template
nop0    ; 110 00  50 copy loop template
movii   ; 110 1a  51 move contents of [bx] to [ax] (copy one instruction)
decC    ; 110 0a  52 decrement cx (size)
ifz     ; 110 05  53 if cx == 0 perform next instruction, otherwise skip it
jmpo    ; 110 14  54 jump to template below (copy procedure exit)
nop0    ; 110 00  55 copy procedure exit complement
nop1    ; 110 01  56 copy procedure exit complement
nop0    ; 110 00  57 copy procedure exit complement
nop0    ; 110 00  58 copy procedure exit complement
incA    ; 110 08  59 increment ax (address in daughter to copy to)
incB    ; 110 09  60 increment bx (address in mother to copy from)
jmpo    ; 110 14  61 bi-directional jump to template below (copy loop)
nop0    ; 100 00  62 copy loop complement
nop1    ; 100 01  63 copy loop complement
nop0    ; 100 00  64 copy loop complement
nop1    ; 100 01  65 copy loop complement
ifz     ; 000 05  66 this is a dummy instruction to separate templates
nop1    ; 110 01  67 copy procedure exit template
nop0    ; 110 00  68 copy procedure exit template
nop1    ; 110 01  69 copy procedure exit template
nop1    ; 110 01  70 copy procedure exit template
popC    ; 110 12  71 pop cx off stack (size)
popB    ; 110 11  72 pop bx off stack (start address of mother)
popA    ; 110 10  73 pop ax off stack (start address of daughter)
ret     ; 110 17  74 return from copy procedure
nop1    ; 100 01  75 end template</code></pre>
<p>Visualization (ALMond):</p>
<blockquote>
<p>In addition, the operating system provides a very elaborate observational system that keeps a record of births and deaths, sequences the code of every creature, and maintains a genebank of successful genomes. The operating system also provides facilities for automating the ecological analysis, that is, for recording the kinds of interactions taking place between creatures. <a href="http://life.ou.edu/tierra/whatis.html">Adapted from &quot;What Tierra Is&quot;</a></p>
</blockquote>
<h3 id="more-information">More information:</h3>
<ul>
<li><a href="http://life.ou.edu/tierra/">The Tierra homepage</a></li>
<li><a href="http://tierra.lolwh.at">Instructions on how to build the Tierra system on a Linux computer</a></li>
<li>A direct descendent of Tierra, and one which has continued to have impact in artificial life research, is <a href="https://en.wikipedia.org/wiki/Avida">Avida</a>. </li>
</ul>
<p><img src="https://upload.wikimedia.org/wikipedia/commons/1/14/Avida_2.6_screenshot.png" alt="Avida"></p>
<!--
http://codepen.io/grrrwaaa/pen/JXXowe?editors=001
-->

<hr>
<h2 id="artificial-chemistries">Artificial chemistries</h2>
<p>One of the most enduring questions facing biology (and artificial life) regards how life itself emerged from non-life, and how that boundary continues to be bridged (&quot;protobiology&quot;). The emergence of life has not yet been observed in physical materials, however there are many hypotheses offered. In general the problem is how life-like behavior and organization emerges from simpler physical chemical systems; or at the very least, how new orders of structure emerge from simpler chemical models. Artificial chemistries (AC) are systems with similarities to models of physics and chemistry which can be used to explore these speculative models with more rigor.</p>
<p>AC&#39;s were proposed by John McCaskill, but probably the best known example (&quot;AlChemy&quot;, described below) was created by Walter Fontana and Leo Buss, presented at the second Artificial Life conference. Hideaki Suzuki and Takashi Ikegami pursued the concept in Japan (with &quot;systems of machines and tapes&quot;, referencing both Turing&#39;s and von Neumann&#39;s early work). <a href="http://www.cs.mun.ca/~banzhaf/papers/alchemistry_review_MIT.pdf">Banzhaf, Dittrich and Ziegler present an excellent review here</a>. See also <a href="http://www.mitpressjournals.org/doi/abs/10.1162/artl.2009.15.1.15100?journalCode=artl">Artificial Chemistry. Hideaki Suzuki, Peter Dittrich. Artificial Life 2009, Vol. 15, No. 1.</a></p>
<p>Like artificial life, in artificial chemistries the assumption holds that organization (information) is the most essential component. One of the central insights stems from a clear analogy between life and computation: organisms are created by organisms, just as programs can create programs. The ease of mapping of code and data mirrors a relationship between physical structure and behavior (form and function). </p>
<p>Artificial chemistries are also one possible attractive option to overcome the &#39;over-specified&#39; nature of most artificial evolutionary and / or developmental systems: a mechanism by which new features come into being, new behaviors, new sensors, new evolutionary pathways, even new modes of evolution. They have also been applied to optimization problems, and to the modeling of social systems. </p>
<h2 id="walter-fontanas-algorithmic-chemistry">Walter Fontana&#39;s Algorithmic Chemistry</h2>
<p>Fontana&#39;s basic conjecture was that adaptive systems are characterized by self-referential processes, in which combinatorial objects encode functions that act back on those same objects. He addressed this by taking two inspirations from chemistry:</p>
<ul>
<li>the collision-reaction of two molecules can create a new molecule</li>
<li>many different reactants can yield the same stable product</li>
</ul>
<p>His insight was to treat fragments of code as equivalent to chemical molecules, and metaprogramming rewrites as equivalent to chemical reactions. That is, code that reacts by modifying itself. Fontana described his system as a &quot;function gas&quot; or &quot;Turing gas&quot;, inspired by chemical flowmreactors, in that a large vessel of code fragments are agitated such that fragments randomly collide, and when they do, one is executed using the other as its input. That is, the code fragments are strings of executable code, that take other strings of code as input, and produce new strings of code as output.  And thus, programs are the material objects of the world.</p>
<blockquote>
<p>This is admittedly somewhat mind-boggling, and gets more so as his papers admit, however perhaps one way to get a grasp on it is to consider how we made an interpreter of a simple character language to generate graphical patterns as we created Biomorphs. Now imagine that rather than generating line segments, we are generating strings of code in the same character language. A crucial difference, which makes this more interesting, is that in Fontana&#39;s work it is not one string that is interpreted, but a pair of strings, where one is being inserted in the other during interpretation. You might observe that this is similar to what our <em>evolutionary</em> process of creating a new generation (rather than the <em>developmental</em> process of interpretation) achieved -- especially in the case of crossover recombination! However an important result of Fontana&#39;s work is that this evolutionary process emerges naturally from the chemistry itself, rather than being imposed from outside.</p>
</blockquote>
<p>Importantly, <strong>there is no control over the order of the chemical reactions</strong> -- code fragments collide at random. The reaction process is thus:</p>
<ul>
<li>select one code fragment at random to execute (i.e. interpet as program)</li>
<li>select another at random to pass in as the input argument (i.e. interpret as data)</li>
<li>the result of the execution is another code fragment</li>
<li>if this code fragment meets some minimal viability conditions (avoiding unreactive or trivial programs), it is added back into the soup, replacing some other randomly chosen code fragment (preserving total population size)</li>
</ul>
<p>This total system showed a remarkable ability to quickly self-organize from populations of random functions to highly specific functions with cooperative characteristics, reliably increasing internal order and complexity. The system showed adaptability to perturbations, and led to a complex set of organizational architectures including catalytic loops, hypercycles, and self-replicating structures.</p>
<blockquote>
<p>&quot;If syntactically legal strings of characters are combinatorial objects and if all legal strings define functions that transform legal strings into lega; strings, then we have an exceedingly simple instance of a combinatorial loop between objects and functions: an algorithmic chemistry, or AlChemy for short.&quot; <a href="http://www.santafe.edu/media/workingpapers/90-011.pdf">Walter Fontana. Algorithmic Chemistry: a model for functional self-organization. 1990</a></p>
</blockquote>
<p>Fontana&#39;s &quot;AlChemy&quot; language was designed for:</p>
<ul>
<li>combinatorial, modular flexibility </li>
<li>lack of syntactic restrictions: that almost any randomly-produced combination of symbols is a legal program</li>
<li>to be unable to cause errors or enter infinite loops</li>
</ul>
<p>It is very simple, consisting of expressions that are either atoms or lists of expressions (including lists of lists of atoms, etc.) Atoms can be one of 11 pre-defined operators, or the symbol &quot;a&quot; to represent the current input variable. </p>
<blockquote>
<p>This language is thus similar to LISP. LISP&#39;s simplicity makes it easy to manipulate through metaprogramming, and it is fairly unique in that data and programs look the same. It builds upon Alonzo Church&#39;s lambda-calculus, which is a functional formulation of computation yet still equivalent to the contemporaneous procedural Turing machine formulation.</p>
</blockquote>
<p>Some example AlChemy molecules:</p>
<pre><code>((-a)a(|(=aaa)a)(-a))
(aaa(%a))
(a(+(+a))(&gt;a)(&lt;a))
(a(%a)a(&#x27;a))
(a(&gt;(=(&gt;a)(-a)(*aa)))(&#x27;(=(*(&lt;a)a)aa))(&lt;(&gt;(!aa))))</code></pre>
<p>The pre-defined operators consume fixed numbers of arguments and return a value result. Other lists produce &quot;functions&quot;, that are applied to the atom immediately to their right (the argument). That is, all the &quot;a&quot; symbols in the function are replaced by the argument, then the function is evaluated to produce its result. The value of a list is defined recursively in terms of the values of its elements. This continues until all elements of the list have been evaluated, typically leaving as a result another list.</p>
<p>Because functions are evaluated and then applied to subsequent arguments there is a high risk of infinite recursions of limitless depth. In AlChemy, expressions are only allowed to recurse to a maximum depth of 10 to prevent this possibility. A time limit was also applied to each interaction.</p>
<h3 id="primitives">Primitives</h3>
<p>The &quot;(&quot; and &quot;)&quot; characters are used to identify lists. </p>
<p>The &quot;a&quot; atom represents the input argument.</p>
<p>&quot;&#39;&quot; this quote operator returns the argument without evaluating it</p>
<p>&quot;%&quot; returns the argument evaluated with the input being null (so long as the depth limit is not reached)</p>
<p>&quot;+&quot; return the first expression of a list (if input is an atom, just return the atom).</p>
<p>&quot;-&quot; return a list with the first expression deleted (if input is an atom, just return the atom).</p>
<p>&quot;&gt;&quot; and &quot;&lt;&quot; are the inverse of &quot;+&quot; and &quot;-&quot;</p>
<p>&quot;*&quot; (2 arguments): if the second arg is a list, then return a copy of this list with the first argument inserted at its head. Otherwise just return the first argument.</p>
<p>&quot;|&quot; (2 arguments): the longer argument is shortened by the length of the shorter argument, and is then returned.</p>
<p>&quot;=&quot; (3 arguments): if first two arguments are identical strings, evaluate &amp; return the third.</p>
<p>&quot;/&quot; (3 arguments): if the length of the first argument is odd, return the second, else the third argument.</p>
<hr>
<p><a href="http://codepen.io/grrrwaaa/pen/YpGGXa?editors=001">Progress toward alchemy</a></p>
<h2 id="further-afield">Further afield</h2>
<p>Here is a <a href="http://chorasimilarity.github.io/chemlambda-gui/dynamic/demos.html">remarkable collection of browser-based demos</a> of the ChemLambda system, which has strong influences from Fontana&#39;s work.</p>
<iframe width="480" height="360" src="https://www.youtube.com/embed/6LZvIkM6JqY?rel=0" frameborder="0" allowfullscreen></iframe>

<h2 id="general-properties-of-artificial-chemistries">General properties of artificial chemistries</h2>
<p>A typical artificial chemistry includes:</p>
<ul>
<li>A set <strong>S</strong> of possible molecule types. This set could be finite or infinite (as far as we know, the set of possible molecules in the universe is infinite.) Molecules could be represented in the program as symbols, character strings, numbers, lambda-expressions, trees, etc.</li>
<li>A set <strong>R</strong> of possible reaction rules, in which input combinations of input molecules produce combinations of output molecules; e.g. <code>A + B -&gt; 2C</code> or <code>A + 2B -&gt; A + D</code>. Note that the second example is a <em>catalytic</em> reaction, since one of the inputs (<code>A</code>) is preserved unchanged. A reaction may also have other conditions, such as ambient temperature; and may have an associated probability (&quot;rate constant&quot;) and energetic cost. </li>
<li>A population <strong>P</strong> of molecules (drawn from <strong>S</strong>) that exist at any moment. </li>
<li>An algorithm <strong>A</strong> specifying how to apply reactions to the population <strong>P</strong> and thus update to the population of the next instant (<strong>P&#39;</strong>), following reaction rules as described below. A reaction may also have other conditions, such as ambient temperature; and may have an associated probability (&quot;rate constant&quot;) and energetic cost. The reaction algorithm could take into account spatial proximity, or it could assume that all chemicals rapidly disperse (the &quot;well-stirred reactor&quot; model), in which reactions are selected stochastically. On the other hand, if concentrations of molecules are represented as continuous numbers, differential equations may be used. </li>
</ul>
<p>An important design choice is how to define the reaction rules. For a finite system these could be designed by hand (or to model a well-known physical system). For an infinite system they must be derived on demand; this could be done programatically according to reactant structures, in an approximation to molecule bonding mechanisms (or protein folding). For example, a simple demonstration AC uses integer division to define reactions: each molecule type is an integer (greater than one), and two molecules can react if they leave no remainder after division (the long-term behavior of this system tends to increase the concentration of prime numbers). Another possibility is to generate reactions and rates randomly as needed. </p>
<p>The algorithm may also include a source and sink: the source is a continuous input of new (base) molecules, and a sink is a gradual removal of molecules to maintain population size. In addition, certain reactions may be filtered to make the whole system more interesting (e.g. discarding reactions that produce nothing more than one of their inputs).</p>
<blockquote>
<p>There are clearly similarities with rewriting systems (where the active population is the item being rewritten, made of molecule <em>terms</em> and reaction <em>productions</em>). There can also be similarities found with evolutionary, cellular and dynamical systems. Alternatively, AC&#39;s can be viewed as a special case of agent systems, in which each agent belongs to a molecule-class, and reactions are a result of agent interacitons (in which interactions usually lead to the creation and destruction of agents). </p>
</blockquote>
<!--

## Genetic Programming

Our math & turtle graphics examples demonstrated how using a programming language as intermediate representation can achieve great power/diversity with succinct syntax. We'll explore that in more detail via *genetic programming*. 

Genetic Programming was invented by Nigel Cramer in 1985, but greatly expanded through the work of John Koza. GP evolves programs; it is an example of **metaprogramming**. GP has been used to generate programs to solve hard problems, and to evolve control systems for artificial agents and robots. Karl Sims used GP for his genetic images as well as for his evolving virtual creatures. 

In GP that the generating a phenotype is a process of generating *programs*, however programs are usually expressed as syntax *trees* rather than as lines of code. The leaves of the tree are *terminals*, while the non-terminal branches are *functions*. Terminals have no inputs; typical terminals are constant numbers and external/global variable names, but this set can also include zero-argument functions such as "random()" or context-sensing such as "get_orientation()". Non-terminal functions are specified according to their operator (such as mathematical addition, multiplication, cosine, etc.); they have one or more inputs (the number of inputs is the *arity*), which may be filled by terminals or other functions. The set of all possible terminals and non-terminal functions constitutes the *primitive set* of a GP system (a union of the *terminal set* and the *function set*). 

This structure is natural to LISP programs:

```lisp
(* (sin (+ x 2)) (cos x))
```

In this case, the terminal set appears to include integers and "x", the function set includes "sin", "cos", and binary operators such as "+", and "*". 

Javascript does not have this representational succinctness, but can get close to it by nested arrays:

```javascript
["*", ["sin", ["+", "x", 2]], ["cos", "x"]]
```

Or more graphically:

```javascript
[
    "*", 
    [
        "sin", 
        [
            "+", 
            "x", 
            2]
    ], 
    [
        "cos", 
        "x"
    ]
]
```

Note that our representation is more brittle than a simple string of symbols: any array beginning with "sin" must have one more element following it as the argument; any array beginning with "+" must have two elements (array length 3) following it as arguments. Moreover, those arguments must either be *terminal* nodes such as "x" or numbers, or they must be arrays -- that is, operator symbols like "sin" and "+" can only occupy the first position. Since we know that our operators require *at most* two arguments, we could simply give all functions two arguments, and ignore the un-used ones as neutral junk genetic data, e.g.:

```javascript
[
    "*", 
    [
        "sin", 
        [
            "+", 
            "x", 
            2], 
        5        // junk
    ], 
    [
        "cos", 
        "x", 
        "x"        // junk
    ]
]
```

### Interpretation (development)

To turn this *data* into something we can actually run as *code*, we could spell it out as an expression:

```javascript
return (Math.sin(x + 2) * Math.cos(x));
```

Here's a recursive solution:

```javascript
function develop(node) {
    // terminals:
    if (typeof node == "string" || typeof node == "number") {
        return node;
    }
    // non-terminals:
    let op = node[0];
    let a = node[1];
    let b = node[2];
    if (op == "sin" || op == "cos") {
        // unary Math function call:
        return "Math." + op + "(" + develop(a) + ")";
    } else if (op == "+" || op == "*" || op == "-" || op == "/" || op == "%") {
        // binary operator:
        return "(" + develop(a) + op + develop(b) + ")";
    }
}
```

### Choice of primitive set

A couple of things to watch out for here: 

- It is assumed that all functions return one value
- It is assumed that returned values are of a type that can be used in other functions. This gets trickier when introducing conditional functions (e.g. `(if x y)`), since numeric values need to be converted to booleans. More complex systems may introduce type metadata into the tree.
- Functions should not throw errors or cause invalid behaviour, so often "protected" versions of operations are used. For example, divide (or modulo) by zero may be detected and avoided, reducing the number of run-time exceptions to trap.
- There should be enough scope in the primitive set to generate satisfactory solutions to the problem

### Seed generation

To generate a random seed genotype, we can also use a recursive procedure. **One thing to watch out for: we need to make sure we don't spawn infinitely large genotypes!** To do that, we'll need to pass some information through the recursion to bail out at a certain depth, or when the genome has a certain size. For example, we could limit the recursive depth as follows:

```javascript
let ops = ["sin", "cos", "+", "-", "*", "/", "%"];

function pick_random_operator() {
    return ops[random(ops.length)];
}
function make_random_terminal() {
    return random(2) ? "time" : Math.random()*1000;
}

function generate(depth) {
    // forces use of terminal when depth == 1, randomized otherwise:
    if (random(depth) == 0) {
        return make_random_terminal();
    } else {
        return [
            pick_random_operator(), 
            generate(depth-1), 
            generate(depth-1)
        ];
    }
}

// generate with maximum depth 6:
var data = generate(6);

// print out nicely:
console.log(JSON.stringify(data, null, " "));
```

Koza also recommends creating a seed population with a variety of depths, which we can easily do by randomizing the initial argument to `generate()`. 


### Mutations and recombinations

GP departs significantly from other evolutionary algorithms in the implementation of mutation and recombination. It is not just individual terminal symbols that can be mutated, but also entire branches. 
Extra care has to be taken here to prune genomes that are getting too large (or to avoid then getting too small!). Some possible mutations include:

- Replace a terminal with another random terminal. A good variation of this is to offset a constant terminal by a small amount.
- Replace the operator name of a function
- Replace a random function with a terminal
- Swap the argument order of a function
- Replace a function with a copy of another function within the same genome (including all subtrees)
- Replace entire genome with one of its subtrees. (This will make it smaller).
- Replace a function with a randomly-generated function (including subtrees). We may introduce a condition to make sure the replacement has a similar size as the item replaced.
- In sexual reproduction, the most common variant is *sub-tree crossover*: Parent A is cloned, but a random subtree is chosen to be replaced with another randomly-chosen subtree from Parent B. Typically sexual cross-over is utilized more when populations number in the hundreds, thousands or more, whereas mutation is more widely utilized in smaller populations.

A tricky point -- most of these mutations depend on choosing a random non-terminal. How can we pick one randomly? And how can we ensure that the genomes stay of a reasonable size? The solutions to these problems are not trivial enough to work through in our labs.


### Alternative representations

Many GP systems today assume that all non-terminal functions have the same arity (the same no. of arguments). By doing so, it turns out that the brackets become redundant, and trees can efficiently be represented as simple linear sequences. The above example can be simply listed as:

```js
* sin + x 2 5 cos x x
```

Clearly this makes some mutations easier -- we can pick random items in the list and replace them with equivalents, e.g. replacing terminals with terminals, operators with operators. We can easily see the genome size, and use that as a selection factor. But it is still not so trivial to make replacements and ensure the whole genome still generates a complete program. 

---

Another variant, called "linear genetic programming", also uses a flat list of nodes, but in a format closer to a register language (in fact, closer to the underlying architecture of modern PCs). Each node in this list has an operator, to determine what kind of instruction it is (e.g. "+", "sin", "var", etc.), followed by some values to specify which existing "registers", that is, previously-computed values, are to be used as arguments to the operation. Each operation will then produce a new register in turn. So, the code generated will look something like this:

```javascript
var r1 = x;
var r2 = 2;
var r3 = r1 + r2;
var r4 = Math.sin(r3);
var r5 = x;
var r6 = Math.cos(r5);
var r7 = r4 * r6;
return r7;
```

A genetic representation might be something like this:

```js
[
    [ "var", "x" ],
    [ "num", 2 ],
    [ "+", 1, 0 ],
    [ "sin", 0 ],
    [ "var", "x" ],
    [ "cos", 0 ],
    [ "*", 2, 0 ],
]
```

Notice how registers are stored as offsets back from the currently-assigning register. So, an argument of `0` means the last-computed-register, `1` means the one computed before that, etc. This kind of relative indexing makes the representation far more robust -- less likely to be malformed after mutation. Even greater safety can be ensured by wrapping register indices within the available range. For a linearized genotype representation, the genetic mutations and crossover operators are similar to other sequence-based evolutionary systems. 

[An example using a kind of linear-GP, somewhat akin Karl Sims' evolving images, is in the lab editor here](http://codepen.io/grrrwaaa/pen/YwmerM?editors=001)

[An example using this scheme to generate random video feedback systems, in the lab editor](http://codepen.io/grrrwaaa/pen/pyzJLe?editors=001)

[An example of this using jit.gen in Max/MSP/Jitter](../code/fb_gen.zip)

---

Yet another very different but related approach is **gramatical evolution**. In GE, genomes are simply lists of integers, for example in the range 0-99. The possible shape of programs is determined by a grammar. The generator follows each rule of the grammar in turn, and when multiple options are possible, the next genome integer is used to determine the choice (modulo the number of choices). This not only guarantees that all generated programs are formally correct (no matter how complex the grammar involved), it also seems to lend some advantages with respect to mutations. Grammatical evolution has been very successful and is widely used.

## Further reading

- [A field guide to GP](http://dces.essex.ac.uk/staff/rpoli/gp-field-guide/A_Field_Guide_to_Genetic_Programming.pdf)
- [An overview paper](http://www.cs.montana.edu/~bwall/cs580/introduction_to_gp.pdf)
- [Very short tutorial](http://www.geneticalgorithms.com/Tutorial/index.html)

-->
<footer>2020</footer>

</body>
</html>